File: playing-cards.py
## In an Amazon interview they asked Poker Hands question
## Write a program that detects flush, two of a kind, three of a kind, full house ect.

flush = ['A1', 'A2', 'A3', 'A4', 'A5']



class Card:
  def __init__(self, card_string):
    self.suit = card_string[0]
    self.value = card_string[1]
  
  def print_card(self):
    print(self.suit + self.value)

class Hand:
  def __init__(self, hand_string):
    self.cards = []
    self.numDict = {}
    self.__create_hand(hand_string)

  def __create_hand(self, hand_string):
    for card in hand_string:
      self.cards.append(Card(card))
      ## create dictionary with numbers
      

  def print_hand(self):
    for card in self.cards:
      card.print_card() 

  def is_flush(self):
    first_suit = None
    for i in range(len(self.cards)):
      if i == 0:
        card = self.cards[i]
        first_suit = card.suit
      else: 
        if first_suit != card.suit:
          return False
    
    return True  

hand_flush = Hand(flush)
hand_flush.print_hand()
print(hand_flush.is_flush())

File: playing-cards.mjs
// In an Amazon interview they asked Poker Hands question
// Write a program that detects flush, two of a kind, three of a kind, full house ect.

class Card {
  constructor(card) {
    this.suit = card[0]
    this.number = card[1]
  }

  printCard() {
    return console.log(this.suit + this.number)
  }
}

class Hand {

  constructor(hand){
    this.cards = hand.map((card) => new Card(card))
    this.suitObj = {}
    this.numObj = {}
    this.#setSuits()
    this.#setNumbers()
  }

  printHand() {
    this.cards.map((card) => card.printCard())
  }
  
  #setSuits() {
    this.cards.map((card) => {
      if (this.suitObj[card.suit]) {
        this.suitObj[card.suit] = this.suitObj[card.suit] + 1
      } else {
        this.suitObj[card.suit] = 1
      }
    })
  }

  #setNumbers() {
    this.cards.map((card) => {
      if (this.numObj[card.number]) {
        this.numObj[card.number] = this.numObj[card.number] + 1
      } else {
        this.numObj[card.number] = 1
      }
    })
  }

  isFlush() {
    for(const property in this.suitObj) {
      if(this.suitObj[property] == 5){
        return true
      }
    }
    return false
  }

  isTwoKind() {
    for(const property in this.numObj) {
      if(this.numObj[property] == 2){
        return true
      }
    }
    return false
  }

}


const flush = ['A1', 'A2', 'A3', 'A4', 'A5']
const twoKind = ['A1', 'B1', 'C2', 'D3', 'F4']

let flushHand = new Hand(flush)
flushHand.printHand
console.log('suitOb', flushHand.suitObj)
console.log('numObj', flushHand.numObj)
console.log('isFlush', flushHand.isFlush())
console.log('isTwoOfAKind', flushHand.isTwoKind())

let twoKindHand = new Hand(twoKind)
console.log('isTwoOfAKind', twoKindHand.isTwoKind())